src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,153 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using System.Collections.Immutable;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.StatelessFuncIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			context.RegisterSyntaxNodeAction(
+				ctx => {
+					if( ctx.Node is ObjectCreationExpressionSyntax expr ) {
+						AnalyzeInvocation( ctx, expr );
+					}
+				},
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			ObjectCreationExpressionSyntax syntax
+		) {
+
+			TypeSyntax type = syntax.Type;
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( type ).Symbol;
+
+			if( symbol == null ) {
+				return;
+			}
+
+			if( !IsStatelessFunc( symbol ) ) {
+				return;
+			}
+
+			ExpressionSyntax argument = syntax.ArgumentList.Arguments[ 0 ].Expression;
```

Comment:
If `ArgumentList` is nullable I'd check that, and I'd check that there is at least one element of `Arguments` before assuming it. Otherwise the analyzer will crash as you type. I don't think it's possible for an arguments `.Expression` to be `null` though.

Reply:
(`ArgumentList` I could buy being `null` depending on how (the invalid code) `new Foo;` gets parsed.)

Reply:
You can add a spec test for something like `new StatelessFunc()` and it should fail because of an exception right now.

Reply:
Ok, cool!